proxy/server/preheat.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2016-2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package server
+
+import (
+	"net/http"
+	_ "net/http/pprof" // Registers /debug/pprof endpoints in http.DefaultServeMux.
+	"github.com/uber/kraken/core"
+	"github.com/uber/kraken/utils/handler"
+	"github.com/uber/kraken/origin/blobclient"
+	"encoding/json"
+	"regexp"
+	"bytes"
+	"github.com/uber/kraken/utils/log"
+	"github.com/uber/kraken/utils/dockerutil"
+	"github.com/uber/kraken/utils/httputil"
+	"time"
+	"fmt"
+	"github.com/docker/distribution"
+)
+
+const manifestPattern = `^application/vnd.docker.distribution.manifest.v\d\+(json|prettyjws)`
+
+// Interface defines the interface of http handler
+type Interface interface {
+	Handle(w http.ResponseWriter, r *http.Request) error
+}
+
+type PreheatHandler struct {
+	clusterClient blobclient.ClusterClient
+}
+
+// NewPreheatHandler creates a new preheat handler.
+func NewPreheatHandler(client blobclient.ClusterClient) Interface {
+	return &PreheatHandler{client}
+}
+
+// Handle notifies origins to cache the blob related to the image
+func (ph *PreheatHandler) Handle(w http.ResponseWriter, r *http.Request) error {
+	var notification Notification
+	if err := json.NewDecoder(r.Body).Decode(&notification); err != nil {
+		return handler.Errorf("decode body: %s", err)
```

Comment:
Add `.Status(http.StatusBadRequest)` so this returns 400.

Reply:
it is hard to say which return code is more applicable， 400 or 500?
for decode error could be caused by an internal server error.
i just follow the same instance in kraken.